geomesa-feature/src/main/scala/org/locationtech/geomesa/feature/FeatureSpecificReader.scala

```diff
@@ -113,6 +111,27 @@ class FeatureSpecificReader(oldType: SimpleFeatureType, newType: SimpleFeatureTy
     sf
   }
 
+  def readWithUserData(reuse: AvroSimpleFeature, in: Decoder): AvroSimpleFeature = {
+    val sf = defaultRead(reuse, in)
+
+    // TODO move out and use everywhere
+    val ar = new AvroReader(in)
+
+    val userData = ar.readGenericMap()
+    sf.getUserData.clear()
+    sf.getUserData.putAll(userData)
+
+    sf
+  }
+
+  private val doRead = {
+    if (opts.withUserData)
+      readWithUserData _
+    else
+      defaultRead _
+  }
+
+  def read(reuse: AvroSimpleFeature, in: Decoder): AvroSimpleFeature = doRead(reuse, in)
```

Comment:
I'm not able to get this to work when overriding an abstract method in a Java class.  For example, in BaseSimpleFeatureSerializer, this does not compile:

  override val write: (Kryo, Output, SimpleFeature) =>  Unit =
    if (opts.withUserData) writeWithUserData else defaultWrite

![selection_007](https://cloud.githubusercontent.com/assets/11392977/7053004/1a69f81e-de00-11e4-877f-2cae920a4eb4.png)